intra/tunnel.go

```diff
@@ -37,95 +37,85 @@ type Listener interface {
 	}
-	if err := t.registerConnectionHandlers(fakedns, dialer, config, listener); err != nil {
-		return nil, err
+
+	fakeDNSAddr, err := net.ResolveUDPAddr("udp", fakedns)
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve fakedns: %w", err)
 	}
-	t.SetDNS(dohdns)
-	return t, nil
-}
 
-// Registers Intra's custom UDP and TCP connection handlers to the tun2socks core.
-func (t *intratunnel) registerConnectionHandlers(fakedns string, dialer *net.Dialer, config *net.ListenConfig, listener Listener) error {
-	// RFC 4787 REQ-5 requires a timeout no shorter than 5 minutes.
-	timeout, _ := time.ParseDuration("5m")
+	t = &Tunnel{
+		sni: &tcpSNIReporter{
+			dns: dohdns,
+		},
+		tun: tun,
+	}
 
-	udpfakedns, err := net.ResolveUDPAddr("udp", fakedns)
+	t.sd, err = makeIntraStreamDialer(fakeDNSAddr.AddrPort(), dohdns, protector, listener, t.sni)
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("failed to create stream dialer: %w", err)
 	}
-	t.udp = NewUDPHandler(*udpfakedns, timeout, config, listener)
-	core.RegisterUDPConnHandler(t.udp)
 
-	tcpfakedns, err := net.ResolveTCPAddr("tcp", fakedns)
+	t.pp, err = makeIntraPacketProxy(fakeDNSAddr.AddrPort(), dohdns, protector, listener)
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("failed to create packet proxy: %w", err)
 	}
-	t.tcp = NewTCPHandler(*tcpfakedns, dialer, listener)
-	core.RegisterTCPConnHandler(t.tcp)
-	return nil
-}
 
-func (t *intratunnel) SetDNS(dns doh.Transport) {
-	t.dns = dns
-	t.udp.SetDNS(dns)
-	t.tcp.SetDNS(dns)
-}
+	if t.IPDevice, err = lwip2transport.ConfigureDevice(t.sd, t.pp); err != nil {
+		return nil, fmt.Errorf("failed to configure lwIP stack: %w", err)
+	}
 
-func (t *intratunnel) GetDNS() doh.Transport {
-	return t.dns
+	t.SetDNS(dohdns)
+	return
 }
 
-func (t *intratunnel) SetAlwaysSplitHTTPS(s bool) {
-	t.tcp.SetAlwaysSplitHTTPS(s)
+// Set the DNSTransport.  This method must be called before connecting the transport
+// to the TUN device.  The transport can be changed at any time during operation, but
+// must not be nil.
+func (t *Tunnel) SetDNS(dns doh.Transport) {
+	t.sd.SetDNS(dns)
+	t.pp.SetDNS(dns)
+	t.sni.SetDNS(dns)
 }
 
-func (t *intratunnel) EnableSNIReporter(filename, suffix, country string) error {
+// Enable reporting of SNIs that resulted in connection failures, using the
+// Choir library for privacy-preserving error reports.  `file` is the path
+// that Choir should use to store its persistent state, `suffix` is the
+// authoritative domain to which reports will be sent, and `country` is a
+// two-letter ISO country code for the user's current location.
+func (t *Tunnel) EnableSNIReporter(filename, suffix, country string) error {
 	f, err := os.OpenFile(filename, os.O_RDWR|os.O_CREATE, 0600)
 	if err != nil {
 		return err
 	}
-	return t.tcp.EnableSNIReporter(f, suffix, strings.ToLower(country))
+	return t.sni.Configure(f, suffix, strings.ToLower(country))
+}
+
+func (t *Tunnel) Disconnect() {
```

Comment:
Delete

Reply:
See the refactor planning comment above.